build-in-services/asset/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+#[cfg(test)]
+mod tests;
+pub mod types;
+
+use bytes::Bytes;
+use derive_more::{Display, From};
+
+use binding_macro::{cycles, init, read, service, write};
+use protocol::traits::{RequestContext, ReturnEmpty, ServiceSDK, StoreMap, RETURN_EMPTY};
+use protocol::types::Hash;
+use protocol::{ProtocolError, ProtocolErrorKind, ProtocolResult};
+
+use crate::types::{
+    Asset, CreateAssetPayload, GetAssetPayload, GetBalancePayload, GetBalanceResponse,
+    TransferPayload,
+};
+
+pub struct AssetService<SDK> {
+    sdk:    SDK,
+    assets: Box<dyn StoreMap<Hash, Asset>>,
+}
+
+#[service]
+impl<SDK: ServiceSDK> AssetService<SDK> {
+    #[init]
+    fn init(mut sdk: SDK) -> ProtocolResult<Self> {
+        let assets: Box<dyn StoreMap<Hash, Asset>> = sdk.alloc_or_recover_map("assets")?;
+
+        Ok(Self { assets, sdk })
+    }
+
+    #[cycles(100_00)]
+    #[read]
+    fn get_asset<Context: RequestContext>(
+        &self,
+        ctx: Context,
+        payload: GetAssetPayload,
+    ) -> ProtocolResult<Asset> {
+        let asset = self.assets.get(&payload.id)?;
+        Ok(asset)
+    }
+
+    #[cycles(100_00)]
+    #[read]
+    fn get_balance<Context: RequestContext>(
+        &self,
+        ctx: Context,
+        payload: GetBalancePayload,
+    ) -> ProtocolResult<GetBalanceResponse> {
+        let balance = self
+            .sdk
+            .get_account_value(&ctx.get_caller(), &payload.asset_id)?
+            .unwrap_or(0);
+        Ok(GetBalanceResponse {
+            asset_id: payload.asset_id,
+            balance,
+        })
+    }
+
+    #[cycles(210_00)]
+    #[write]
+    fn create_asset<Context: RequestContext>(
+        &mut self,
+        ctx: Context,
+        payload: CreateAssetPayload,
+    ) -> ProtocolResult<Asset> {
+        let caller = ctx.get_caller();
+        let payload_str =
+            serde_json::to_string(&payload).map_err(|e| ServiceError::JsonParse(e))?;
+
+        let id = Hash::digest(Bytes::from(payload_str + &caller.as_hex()));
+
+        if self.assets.contains(&id)? {
```

Comment:
No symbol, name check?

Reply:
Id is generated by caller + name + symbol